### PR TITLE
Feature/785 documentation metadata

### PR DESCRIPTION
**Pull request by JulianFeinauer**
This PR adds Documentation for the current implementation of the Metadata Feature and the necessary adjustment to the Swagger docs.

### CODE REVIEW

**Comment by w4tsn** [comment]
The [json-schema v2](https://github.com/eclipse/ditto/blob/58c5c4ad20fcc4385ae8f743e085de4592ec2efd/documentation/src/main/resources/jsonschema/thing_v2.json) needs to be changed too

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+## Setting custom Metadata
+
+The only way to set arbitrary `Metadata` is by using the `put-metadata` header, see [here](protocol-specification.html#headers) for more information.
+The Format of the Json Array is 
+
+```
+[
+  {
+    "key":"/features/lamp/properties/color/r",
+    "value":{"issuedAt":someTimestamp,"issuedBy":{"name":"me","mail":"me@mail.com"}}
```
Comment (w4tsn) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:37]: Is that really possible? According to https://github.com/eclipse/ditto/issues/785#issuecomment-688062943 I'd assume that each metadata should receive it's own array entry?

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+will yield the Metadata stored for the given Thing, in the following format:
+
+```json
+{
+    "thingId": ...,
+    "attributes": {
+        ...
+    },
+    "_metadata": {
+        ...
```
Comment (w4tsn) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:79]: I'd use a full example here so a user may see the whole thing in action like this (from https://github.com/eclipse/ditto/issues/778#issue-687274224):

```json
{
  "thingId": "org.eclipse.ditto:thing-1",
  "policyId": "...",
  "features": {
    "lamp": {
      "properties": {
        "on": true,
        "color": {
          "r": 0,
          "g": 255,          
          "b": 255,
        }
      }
    }
  },
  "_created": "2020-06-01T10:00:00Z",
  "_modified": "2020-06-09T14:30:00Z",
  "_revision": 42,
  "_metadata": {
    "features": {
      "lamp": {
        "properties": {
          "on": {
            "issuedAt": "2020-06-09T14:30:00Z"
          },
          "color": {
            "r": {
              "issuedAt": "2020-06-09T14:15:00Z"
            },
            "g": {
              "issuedAt": "2020-06-09T14:15:00Z"
            },
            "b": {
              "issuedAt": "2020-06-09T14:15:00Z"
            }
          }
        }
      }
    }
  }
}
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    "key":"*/timestamp",
+    "value": someTimestamp
+  }
+]
+```
+
+## Reading Metadata Information
+
+Currently, the only way to retrieve stored Metadata is by a full thing query, e.g. via the [HTTP API](http-api-doc.html).
+But, `get-metadata` has to be added to the queried fields together with `thingId,attributes,_metadata`.
```
Comment (w4tsn) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:67]: What's that about `get-metadata`? Also this sentence seems to be incomplete

File: documentation/src/main/resources/pages/ditto/protocol-specification.md [hunk 4]
```diff
@@ -78,6 +78,7 @@ There are some pre-defined headers, which have a special meaning for Ditto:
 | `requested-acks` | Defines which [acknowledgements](basic-acknowledgements.html) are requested for a command processed by Ditto. | `JsonArray` of `String` - default: `["twin-persisted"]` |
 | `timeout` | Defines how long the Ditto server should wait, e.g. applied when waiting for requested acknowledgements. | `String` - e.g.: `42s` or `250ms` or `1m` - default: `60s`|
 | `version` | Determines in which schema version the `payload` should be interpreted. | `Number` - currently: \[1,2\] |
+| `metadata` | Determines which metadata information is stored in the thing. | `JsonArray` of `String` - default: `` |
```
Comment (w4tsn) [inline documentation/src/main/resources/pages/ditto/protocol-specification.md:81]: Is the data type correct? It's actually JsonArray of at least Object? Maybe MetadataEntry? It's a complex datatype after all

**Review by w4tsn — CHANGES_REQUESTED** [review]
Added some questions and suggestions.

File: documentation/src/main/resources/pages/ditto/basic-metadata.md (hunk 1 above)
Comment (JulianFeinauer) [reply documentation/src/main/resources/pages/ditto/basic-metadata.md:37]: Hey, I'm refering to https://github.com/eclipse/ditto/issues/680#issuecomment-675446480 here. I think I tested it but have to retest it to be sure.

File: documentation/src/main/resources/pages/ditto/basic-metadata.md (hunk 2 above)
Comment (JulianFeinauer) [reply documentation/src/main/resources/pages/ditto/basic-metadata.md:79]: thanks, I will add that : )

File: documentation/src/main/resources/pages/ditto/protocol-specification.md (hunk 4 above)
Comment (JulianFeinauer) [reply documentation/src/main/resources/pages/ditto/protocol-specification.md:81]: You are absolutely right. @thjaeckle can you comment on the exact type. I have to few knowledge about dittos type system :/

File: documentation/src/main/resources/pages/ditto/basic-metadata.md (hunk 3 above)
Comment (thjaeckle) [reply documentation/src/main/resources/pages/ditto/basic-metadata.md:67]: get-metadata is not yet implemented and not required to be set currently

File: documentation/src/main/resources/pages/ditto/protocol-specification.md (hunk 4 above)
Comment (thjaeckle) [reply documentation/src/main/resources/pages/ditto/protocol-specification.md:81]: First, "metadata" is the wrong header key. It must be "put-metadata".
The data type is JsonArray of JsonObjects containing a "key" with the value being a JsonString and a "value" with the value being a JsonValue (any valid json value should work here, also arrays, objects).

File: documentation/src/main/resources/pages/ditto/basic-metadata.md (hunk 3 above)
Comment (JulianFeinauer) [reply documentation/src/main/resources/pages/ditto/basic-metadata.md:67]: sorry, indeed. Typo.

File: documentation/src/main/resources/openapi/ditto-api-2.yml [hunk 5]
```diff
@@ -4858,6 +4863,14 @@ components:
...
       schema:
         type: string
+    putMetadataParam:
+      name: Put-Metadata
+      in: header
+      description:  >-
+        The `put-metadata` header, which sets Metadata information in the Thing.
+      required: false
+      schema:
+        type: string
```
Comment (thjaeckle) [inline documentation/src/main/resources/openapi/ditto-api-2.yml:4873]:
```suggestion
        type: array
        description: An array of objects containing metadata to apply.
        items:
          type: object
          description: Object containing a `key` where to apply the metadata and a `value` with the metadata value to apply.
          additionalProperties:
            type: object
              properties:
                key:
                  type: string
                  description: The JsonPointer to set the metadata `value` to. May start with `*/<key>` in order to apply the metadata to all affected JSON leafs.
                value:
                  description: The arbitrary JSON value to set as metadata.
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
+---
+title: Metadata Support
+keywords: metadata, things
+tags: [model]
+permalink: basic-metadata.html
+---
+
+Besides [features](basic-feature.html) a Thing in Ditto is also able to store Metadata information. 
+These informations can be additional information for describing specific features or properties of the thing.
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:9]:
```suggestion
These information can be additional information for describing specific features or properties of the thing.
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+A very simple example is the timestamp when the current value of the feature or property was set.
+
+Metadata can be manipulated by using the `metadata` header, see [here](protocol-specification.html#headers) for more information.
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:12]:
```suggestion
Metadata can be manipulated by using the `put-metadata` header, see [here](protocol-specification.html#headers) for more information.
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+<!--
+Right now, the only fixed pre-defined Metadata key is `issuedAt`.
+
+### The `issuedAt` Metadata key
+ 
+The Metadata key `issuedAt` describes the timestamp when the current value of a feature property was _recorded_ (_not when it was set_).
+This means, that a client _can_ send this value by using the `metadata` header. If the client does not send
+this header the `issuedAt` field will be set by Ditto with the timestamp when the value was written.
+-->
+
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:27]:
```suggestion
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+## Setting custom Metadata
+
+The only way to set arbitrary `Metadata` is by using the `put-metadata` header, see [here](protocol-specification.html#headers) for more information.
+The Format of the Json Array is 
+
+```
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:33]:
```suggestion
```json
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md (hunk 1 above)
Comment (thjaeckle) [reply documentation/src/main/resources/pages/ditto/basic-metadata.md:37]: should be, yes - you set an JsonObject for the `"/features/lamp/properties/color/r"` property.

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ 
+The Metadata key `issuedAt` describes the timestamp when the current value of a feature property was _recorded_ (_not when it was set_).
+This means, that a client _can_ send this value by using the `metadata` header. If the client does not send
+this header the `issuedAt` field will be set by Ditto with the timestamp when the value was written.
+-->
+
+## Setting custom Metadata
+
+The only way to set arbitrary `Metadata` is by using the `put-metadata` header, see [here](protocol-specification.html#headers) for more information.
+The Format of the Json Array is
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:31]:
```suggestion
The format of the JSON array is 
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+]
+```
+
+where `key` describes the hierarchical position in the Thing where the metadata is placed and 
+`value` is a map of Metadata keys to their respective values.
+
+A special Syntax for the key is `*/{key}` which means that all affected Json Leafs of the Modify Operation will
+get the Metadata Key `{key}` with the given value. So if, for example, only the affected Json Leafes should 
+get the timestamp where the changed values were recorded, one would set the `metadata` header as shown in the
+following example:
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:53]:
```suggestion
A special syntax for the key is `*/{key}` which means that all affected JSON leafs of the modify operation will
get the Metadata key `{key}` with the given value. So if, for example, only the affected JSON leafs should 
get the timestamp where the changed values were recorded, one would set the `put-metadata` header as shown in the following example: 
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    "key":"*/timestamp",
+    "value": someTimestamp
+  }
+]
+```
+
+## Reading Metadata Information
+
+Currently, the only way to retrieve stored Metadata is by a full thing query, e.g. via the [HTTP API](http-api-doc.html).
+But, `_metadata` has to be added to the queried fields together with `thingId,attributes,_metadata`.
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:67]:
```suggestion
Currently the only way to retrieve stored Metadata is by querying a full thing, e.g. via the [HTTP API](http-api-doc.html), and specifying an (additional) [field selector](httpapi-concepts.html#with-field-selector) `_metadata`, e.g.: `?fields=thingId,attributes,_metadata`.
```

File: documentation/src/main/resources/pages/ditto/basic-metadata.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+For example a `GET` request to `https://{ditto-instance}/api/2/things/{namespace}:{thing}?fields=thingId,attributes,_metadata`
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/basic-metadata.md:69]:
```suggestion
For example a `GET` request to `https://{ditto-instance}/api/2/things/{namespace}:{name}?fields=thingId,policyId,features,_created,_modified,_revision,_metadata`
```

File: documentation/src/main/resources/pages/ditto/protocol-specification.md [hunk 14]
```diff
@@ -78,6 +78,7 @@ There are some pre-defined headers, which have a special meaning for Ditto:
 | `requested-acks` | Defines which [acknowledgements](basic-acknowledgements.html) are requested for a command processed by Ditto. | `JsonArray` of `String` - default: `["twin-persisted"]` |
 | `timeout` | Defines how long the Ditto server should wait, e.g. applied when waiting for requested acknowledgements. | `String` - e.g.: `42s` or `250ms` or `1m` - default: `60s`|
 | `version` | Determines in which schema version the `payload` should be interpreted. | `Number` - currently: \[1,2\] |
+| `put-metadata` | Determines which Metadata information is stored in the thing. | `JsonArray` of `JsonObject`s - default: `[]` |
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/protocol-specification.md:81]:
```suggestion
| `put-metadata` | Determines which Metadata information is stored in the thing. | `JsonArray` of `JsonObject`s containing [metadata](basic-metadata.html) to apply. |
```

File: documentation/src/main/resources/pages/ditto/protocol-specification.md [hunk 15]
```diff
@@ -97,6 +98,8 @@ HTTP protocol, for example the prefix `ditto-*`.
 The interaction between the headers `response-required`, `requested-acks` and `timeout` is documented
 [here](basic-acknowledgements.html#interaction-between-headers).
 
+More information of the Metadata management of Ditto and predefined Metadata-Keys can be found [here](basic-metadata.html).
```
Comment (thjaeckle) [inline documentation/src/main/resources/pages/ditto/protocol-specification.md:101]:
```suggestion
```

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Some change requests.

And what is still missing: adding the new `basic-metadata.html` page to the navigation. Currently, it won't show up in the page navigation at all.

**Review by thjaeckle — APPROVED** [review]

**Comment by thjaeckle** [comment]
@JulianFeinauer could you have a look on the changes I made? Are they ok for you?
I really want to get the docs merged - it's not good to have no docs for the metadata feature for such a long time ;)

**Comment by JulianFeinauer** [comment]
THanks for your support here, all changes are fine for me 👍

**Pull request merged**